### PR TITLE
Add testinfra modules

### PR DESCRIPTION
**Pull request by blarghmatey**
### What does this PR do?

Add modules for using Testinfra in salt states and from the salt CLI
### What issues does this PR fix or reference?

None
### New Behavior

It is now possible to write unit tests for minion state using the functionality contained in the Testinfra library.
### Tests written?

Not yet

### CODE REVIEW

File: salt/modules/testinframod.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+import types
+
+log = logging.getLogger(__name__)
+
+try:
+    import testinfra
+    from testinfra import modules
+    TESTINFRA_PRESENT = True
+except ImportError:
+    log.debug('Unable to import TestInfra')
```
Comment (s0undt3ch) [inline salt/modules/testinframod.py]: Instead of logging this here...

File: salt/modules/testinframod.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+
+
+class InvalidArgumentError(Exception):
+    pass
+
+
+def __virtual__():
+    if TESTINFRA_PRESENT:
+        return __virtualname__
+    return False
```
Comment (s0undt3ch) [inline salt/modules/testinframod.py]: Do:

```python
return False, "The Testinfra package is not available"
```

File: salt/states/testinframod.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import re
+import logging
+
+log = logging.getLogger(__name__)
+
+try:
+    from testinfra import modules
+    TESTINFRA_PRESENT = True
+except ImportError:
+    log.debug('Unable to import TestInfra')
```
Comment (s0undt3ch) [inline salt/states/testinframod.py]: The same as on the module...

**Comment by cachedout** [comment]
Really cool! I like this a lot, @blarghmatey. Thanks very much for this.

**Comment by blarghmatey** [comment]
Thanks @cachedout. I enjoyed writing it and I have been having good luck using it for my formulas which you can review at https://github.com/mitodl/?utf8=✓&query=-formula

One thing that I haven't done yet but would like to do is to automatically generate the doc strings for each exposed function based on the docstrings and methods that are present in the corresponding Testinfra module. For example, the docstring for the `Service` module would reflect what is found at http://testinfra.readthedocs.io/en/latest/modules.html#service

Another contribution that may be useful is to wrap the testinfra execution module with a runner.

**Comment by cachedout** [comment]
@blarghmatey 

Just FYI, this is failing on Python 3. We'll need to find a way to fix this soon or we'll need to revert this since we're now moving very quickly toward full Python 3 support. Here is the stacktrace:

```
[ERROR   ] Failed to import module testinframod, this is due most likely to a syntax error:
Traceback (most recent call last):
  File "/home/mp/devel/salt/salt/loader.py", line 1309, in _load_module
    ), fn_, fpath, desc)
  File "/usr/lib/python3.5/imp.py", line 234, in load_module
    return load_source(name, filename, file)
  File "/usr/lib/python3.5/imp.py", line 172, in load_source
    module = _load(spec)
  File "<frozen importlib._bootstrap>", line 693, in _load
  File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 662, in exec_module
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
  File "/home/mp/devel/salt/salt/modules/testinframod.py", line 295, in <module>
    _register_functions()
  File "/home/mp/devel/salt/salt/modules/testinframod.py", line 287, in _register_functions
    for module_ in modules.__all__:
NameError: name 'modules' is not defined
```

**Comment by blarghmatey** [comment]
@cachedout Thank you for bringing that to my attention. I have added a fix in the PR referenced above.

**Pull request merged**